### DISCUSSION TITLE
Port to Typescript

### DISCUSSION THREAD

**Discussion by wwerner** [gr3p1p3/transparent-proxy#28] [Ideas]
Hey there,

we're planning to use a proxy built on top of this in production, and I'm kinda afraid of maintaining it. There are several places that use dynamic types that are not easy to understand (at least for me) when looking at the code. In my opinion, being more strict with typing would go a long way in alleviating that.

I've made an initial PoC port [here](https://github.com/wwerner/transparent-proxy/pull/2), seems to be not too hard.
After that, I assume future refactorings/maintenance work to be safer and better understandable.

Examples I personally consider worth improving in that regard:
* Checking `isFunction` at runtime. The compiler could do that for us. https://github.com/wwerner/transparent-proxy/blob/chore/typescript/src/lib/usingUpstreamToProxy.ts#L8
* `Session.response` can be `Buffer`, `HttpMessage` or `undefined`, which forces us to deal with all possibilities in the code. This is up to the developer and not guided by the IDE/Linter/Compiler. I think it's easy to miss handling all options when changing things and it might blow up only at runtime. In combination with using a library for parsing http messages #23, I imagine the code dealing with messages to become significantly simpler and harder to break accidentally. https://github.com/wwerner/transparent-proxy/blob/chore/typescript/src/core/Session.ts#L130 

I'd gladly help out making this happen, if you're interested. 
WDYT?

**Comment by gr3p1p3** [comment]
I think the first point is a good point and I'm thinking about putting this on roadmap.

In the v1.11.7 `Session.response` is an object that will be empty if the connection is encrypted.
Honestly, I'm thinking about parsing the HTTP-Payloads myself and doing a big refactor :)
